src/VisualStudio/Core/Def/Implementation/ProjectSystem/CPS/ProjectContextFactory.ProjectContext_IProjectContext.cs

```diff
@@ -0,0 +1,95 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
+{
+    internal partial class ProjectContextFactory
+    {
+        private sealed partial class ProjectContext : AbstractProject, IProjectContext
+        {
+            #region Options
+            void IProjectContext.SetCommandLineArguments(CommandLineArguments commandLineArguments)
+            {
+                base.SetCommandLineArguments(commandLineArguments);
+            }
+            #endregion
+
+            #region References
+            public void AddMetadataReference(string referencePath, MetadataReferenceProperties properties)
+            {
+                referencePath = FileUtilities.NormalizeAbsolutePath(referencePath);
+                AddMetadataReferenceAndTryConvertingToProjectReferenceIfPossible(referencePath, properties);
+            }
+
+            void IProjectContext.RemoveMetadataReference(string referencePath)
+            {
+                referencePath = FileUtilities.NormalizeAbsolutePath(referencePath);
+                base.RemoveMetadataReference(referencePath);
+            }
+
+            public void AddProjectReference(ProjectId projectId, MetadataReferenceProperties properties)
+            {
+                var referencedProject = this.ProjectTracker.GetProject(projectId);
+                var binPathOpt = referencedProject.TryGetBinOutputPath();
+                if (!string.IsNullOrEmpty(binPathOpt))
```

Comment:
AbstractProject and ProjectTracker is tracking project references through bin paths. I'll add a comment.